graph/encoding/dot: add multigraph serialization and deserialization support

32bitkid opened the pull request:
This should resolve issue #396, in the way that has been discussed there. This change introduces a new pair of `MarshalMulti()` and `UnmarshalMulti()` functions, both which take `graph.Multigraph`. Please take a look.

There are a couple outstanding questions on this particular implementation:

- The arguments for `Marshal()` have remained unchanged. However, the new function `MarshalMulti()` does not include the `strict` argument, because it doesn't _really_ make any sense in that context. But, I wonder if `Marshal()` should _always_ emit a strict DOT graph, because thats what it really represents. However, that might be too big of a breaking change here.

-  In the `multiprinter.print()` function, I need to iterate lines, which I do via `lines := graph.LinesOf(g.Lines(nid, tid))`, I'm not sure if I need to _explicitly_ sort the lines here to get a stable output, like one does a few lines up on `NodesOf()`, and if so, then what would be the preferred sorting strategy to use here?

If the approach looks worth moving forward, then I'll fill in some more testing coverage before finalizing the PR.

Thanks!

<!--
Checklist:

- API changes have been discussed
- code is goformated correctly (goimports)
- packages with generated code have had code generation run
- tests pass locally
- linked to relevant issues

Please make sure your commit message summary line and pull request
title match the Go convention; a one-line summary of the change,
prefixed by the primary affected package that should complete the
sentence, "This change modifies Gonum to _____."
-->

kortschak commented:
Addressing your questions before I take a look.

There are reasons why you might not want `strict` on a simple graph and why you might want `strict` on a DOT rendering of a multigraph (in this latter case, you have a multigraph internally that has had some work done on it that you expect to leave it as a simple graph in the multigraph container). Whether this check is work that should happen in the graph as opposed to the serialisation is not immediately obvious to me. The other option is for neither serialisation routine to prepend with `strict ` and do as the xml package does and let the user just do `"strict %s"` or similar. I think this warrants further discussion, and I'm not particularly worried about breaking things if we get it right.

The lines will need to be ordered. The appropriate ordering would be lexical on DOTID with the from node being the first key and the to node being the second key.

32bitkid commented:
Re `strict`: I think the solution that I would, personally, be most in favor of would be to remove the `strict` parameter, and be to have the `Marshal(graph.Graph)` method _always_ output a `strict graph` and `MarshalMulti(graph.Multigraph)` _never_ output a `strict` and drop the parameter for both, and then introduce a new method for converting between `graph.Graph` and `graph.Multigraph` where it would return an `error` if it is not convertible. But thats just me, and again, I feel like its beyond the scope of this PR specifically.

As such, I think your opinion on why one might serialize a `strict` multigraph is well-taken. For the short-term, I think it's probably best to re-introduce the argument here, and then the signatures and functionality will be the same for both. And a more thoughtful discussion on what the right path is forward can result in another PR.

On the latter, I'll make some updates and tests to verify the ordering is correct and stable; I agree they need to be stable, just wasn't sure if it needed _explicit_ sorting in the MarshalMulti() loop, or if just asking for `g.Lines(from, to)` was enough to guarantee correct ordering.

kortschak commented:
Thanks.

In the ordering of things like lines between a pair of node, I go out of my way to not guarantee any ordering unless it makes sense to have an ordering (like here). The rationale for this is that having an ordering on items that are not intrinsically ordered leads to bad graph algorithm implementations.

kortschak commented:
BTW Would you file an issue to look at the strict situation here. Add the <kbd>stability</kbd> label.

32bitkid commented:
@kortschak I created an issue, and described it the best I can. #625 However, as a contributor, I don't think I'm privileged enough to assign labels. Feel free to reword it or change it however you see fit. Thanks!

kortschak commented:
Is this ready for review?

32bitkid commented:
@kortschak i think this is ready for review now. Thanks!

kortschak commented:
Thanks. I'll be able to get to this in a few days.

kortschak commented on graph/encoding/dot/decode.go:
```diff
@@ -50,6 +50,19 @@ func Unmarshal(data []byte, dst encoding.Builder) error {
 }
 
+// UnmarshalMulti parses the Graphviz DOT-encoded data as a multigraph and
+// stores the result in dst.
```
We should probably have a note that only the first graph of one can be stored (same for `Unmarshal`).

I'll file an issue.

kortschak commented on graph/encoding/dot/decode.go:
```diff
@@ -117,8 +162,85 @@ func (gen *generator) node(dst encoding.Builder, id string) graph.Node {
 	return n
 }
 
+// applyPortsToEdge applies the available port metadata from an ast.Edge
```
Can you reorder this to reduce diff noise please.

kortschak commented on graph/internal/ordered/sort.go:
```diff
@@ -74,3 +74,20 @@ func Reverse(nodes []graph.Node) {
 	}
 }
+
+// Lines implements the sort.Interface sorting a slice of graph.Line
```
Please rename this to `LinesByIDs`.

32bitkid commented:
Will do.

> On Nov 13, 2018, at 5:58 PM, Dan Kortschak <notifications@github.com> wrote:
> 
> @kortschak commented on this pull request.
> 
> Do you reckon you could rebase this onto master and force push. I'd like to review without the merge from master noise.
> 
> Thanks
> 
> In graph/encoding/dot/decode.go:
> 
> > @@ -50,6 +50,19 @@ func Unmarshal(data []byte, dst encoding.Builder) error {
>  	return copyGraph(dst, file.Graphs[0])
>  }
>  
> +// UnmarshalMulti parses the Graphviz DOT-encoded data as a multigraph and
> +// stores the result in dst.
> We should probably have a note that only the first graph of one can be stored (same for Unmarshal).
> 
> I'll file an issue.
> 
> In graph/encoding/dot/decode.go:
> 
> > @@ -117,8 +162,85 @@ func (gen *generator) node(dst encoding.Builder, id string) graph.Node {
>  	return n
>  }
>  
> +// applyPortsToEdge applies the available port metadata from an ast.Edge
> Can you reorder this to reduce diff noise please.
> 
> In graph/internal/ordered/sort.go:
> 
> > @@ -74,3 +74,20 @@ func Reverse(nodes []graph.Node) {
>  		nodes[i], nodes[j] = nodes[j], nodes[i]
>  	}
>  }
> +
> +// Lines implements the sort.Interface sorting a slice of graph.Line
> Please rename this to LinesByIDs.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

kortschak commented:
See @mewmew's comment on self-loop test cases [here](https://github.com/gonum/gonum/issues/648#issuecomment-438578846).

32bitkid commented:
@kortschak I rebased and reorganized the code a bit to minimize the diff noise. I also added some tests/assertions for multigraph self-loops.

Though, while the re-organization of the code makes the diff a little easier to parse/approve this PR, it probably need to be _actually_ be re-ogranized into a more idiomatic form after with proper method grouping.

kortschak commented on graph/encoding/dot/encode.go:
```diff
@@ -265,11 +282,44 @@ func (p *printer) print(g graph.Graph, name string, needsIndent, isSubgraph bool
 	return nil
 }
 
+func (p *printer) printBackMatter() {
```
Do we need this?

32bitkid replied:
I went back and forth on this; and I'm not sure... It felt weird to _not_ have it given that it starts with a common printFrontMatter(), it felt unsymmetrical to unroll it and print a seemingly unmatched `}` here. But, as I think you are implying it feels superfluous to print a single `}`. So, I'm not sure and I'll defer to whatever you think is best.

kortschak replied:
Can you remove it.

kortschak commented on graph/encoding/encoding.go:
```diff
@@ -12,6 +12,12 @@ type Builder interface {
 	graph.Builder
 }
 
+// Builder is a graph that can have user-defined nodes and edges added.
```
`MultiBuilder`

32bitkid commented on graph/encoding/dot/decode.go:
```diff
@@ -117,8 +162,10 @@ func (gen *generator) node(dst encoding.Builder, id string) graph.Node {
 	return n
 }
 
+type graphGenerator struct{ generator }
```
I'm okay with this, especially because they are package private.

kortschak commented:
Thanks for doing this.

32bitkid commented:
Happy to help. Thanks to everybody for all the work they’ve put in. 

> On Nov 24, 2018, at 3:02 PM, Dan Kortschak <notifications@github.com> wrote:
> 
> Thanks for doing this.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
